File: sdks/csharp/src/Sudomimus.Token/TokenVerifier.cs
namespace Sudomimus.Token;

/// <summary>
/// Resolves an application's PEM-encoded RSA public key from its anchor.
/// Mirrors <c>@sudomimus/token</c>'s <c>PublicKeyResolver</c>.
/// </summary>
/// <param name="applicationAnchor">
/// The token's audience claim — typically the issuing application's anchor.
/// </param>
/// <param name="keyId">The JWT JOSE header <c>kid</c> used to select a JWK.</param>
/// <param name="cancellationToken">Token observed for cooperative cancellation.</param>
public delegate Task<string> PublicKeyResolver(
    string applicationAnchor,
    string keyId,
    CancellationToken cancellationToken);

/// <summary>
/// Verifies Sudomimus access and refresh tokens end-to-end: structural
/// integrity, expected token media type, audience presence, expiration, and RSA
/// signature against a caller-supplied public key.
/// </summary>
public sealed class TokenVerifier
{
    public const string AccessTokenType = "vnd.sudomimus.application-access+jwt";
    public const string RefreshTokenType = "vnd.sudomimus.application-refresh+jwt";

    private readonly PublicKeyResolver _resolver;
    private readonly Func<DateTimeOffset> _clock;

    public TokenVerifier(PublicKeyResolver resolver)
        : this(resolver, () => DateTimeOffset.UtcNow)
    {
    }

    /// <param name="resolver">Resolver for the application's PEM public key.</param>
    /// <param name="clock">Override "now" for tests. Defaults to <c>DateTimeOffset.UtcNow</c>.</param>
    public TokenVerifier(PublicKeyResolver resolver, Func<DateTimeOffset> clock)
    {
        _resolver = resolver;
        _clock = clock;
    }

    /// <summary>
    /// Parse, verify, and return a Sudomimus access token. Throws
    /// <see cref="TokenException"/> with a categorized code on any failure.
    /// </summary>
    public Task<JwtToken<AccessTokenBody>> VerifyAccessTokenAsync(string jwt, CancellationToken ct = default)
        => VerifyAsync(jwt, AccessTokenType, TokenParser.ParseAccessToken, ct);

    /// <summary>
    /// Parse, verify, and return a Sudomimus refresh token.
    /// </summary>
    public Task<JwtToken<RefreshTokenBody>> VerifyRefreshTokenAsync(string jwt, CancellationToken ct = default)
        => VerifyAsync(jwt, RefreshTokenType, TokenParser.ParseRefreshToken, ct);

    private async Task<JwtToken<TBody>> VerifyAsync<TBody>(
        string jwt,
        string expectedTokenType,
        Func<string, JwtToken<TBody>> parser,
        CancellationToken ct)
        where TBody : class
    {
        // Peek first so a wrong-type token surfaces as WrongTokenType rather
        // than InvalidJwt from the expected payload shape.
        var peeked = TokenParser.PeekHeader(jwt);
        if (!string.Equals(peeked.Type, expectedTokenType, StringComparison.Ordinal))
        {
            throw new TokenException(
                TokenErrorCode.WrongTokenType,
                $"Expected token type \"{expectedTokenType}\", got \"{peeked.Type}\".");
        }

        var payload = TokenParser.PeekBody(jwt);
        var audience = payload.TryGetProperty("aud", out var audienceElement)
            && audienceElement.ValueKind == System.Text.Json.JsonValueKind.String
            ? audienceElement.GetString()
            : null;
        if (string.IsNullOrEmpty(audience))
        {
            throw new TokenException(
                TokenErrorCode.MissingAudience,
                "Token is missing the `aud` (applicationAnchor) payload claim.");
        }

        var keyId = peeked.KeyId;
        if (string.IsNullOrEmpty(keyId))
        {
            throw new TokenException(
                TokenErrorCode.MissingKeyId,
                "Token is missing the `kid` signing-key identifier.");
        }

        var parsed = parser(jwt);

        if (!parsed.VerifyExpiration(_clock()))
        {
            throw new TokenException(TokenErrorCode.Expired, "Token has expired.");
        }

        var publicKey = await _resolver(audience, keyId, ct).ConfigureAwait(false);

        if (!parsed.VerifySignature(publicKey))
        {
            throw new TokenException(
                TokenErrorCode.InvalidSignature,
                "Token signature does not match the application public key.");
        }

        return parsed;
    }
}

File: sdks/java/token/src/main/java/com/sudomimus/token/TokenVerifier.java
package com.sudomimus.token;

import java.time.Clock;
import java.time.Instant;
import java.util.Objects;
import java.util.function.Function;

/**
 * Verifies Sudomimus access and refresh tokens end-to-end: structural
 * integrity, expected token media type, audience and key ID presence, expiration, and
 * RSA signature against a caller-supplied public key.
 */
public final class TokenVerifier {

    public static final String ACCESS_TOKEN_TYPE = "vnd.sudomimus.application-access+jwt";
    public static final String REFRESH_TOKEN_TYPE = "vnd.sudomimus.application-refresh+jwt";

    private final PublicKeyResolver resolver;
    private final Clock clock;

    public TokenVerifier(PublicKeyResolver resolver) {
        this(resolver, Clock.systemUTC());
    }

    /**
     * @param resolver resolver for the application's PEM public key.
     * @param clock    override "now" for tests. Defaults to {@link Clock#systemUTC()}.
     */
    public TokenVerifier(PublicKeyResolver resolver, Clock clock) {
        this.resolver = Objects.requireNonNull(resolver, "resolver");
        this.clock = Objects.requireNonNull(clock, "clock");
    }

    public JwtToken<AccessTokenBody> verifyAccessToken(String jwt) {
        return verify(jwt, ACCESS_TOKEN_TYPE, TokenParser::parseAccessToken);
    }

    public JwtToken<RefreshTokenBody> verifyRefreshToken(String jwt) {
        return verify(jwt, REFRESH_TOKEN_TYPE, TokenParser::parseRefreshToken);
    }

    private <TBody> JwtToken<TBody> verify(
            String jwt,
            String expectedTokenType,
            Function<String, JwtToken<TBody>> parser) {

        // Peek header first so a wrong-type token surfaces as WRONG_TOKEN_TYPE.
        JwtHeader peeked = TokenParser.peekHeader(jwt);
        if (!expectedTokenType.equals(peeked.type)) {
            throw new TokenException(TokenErrorCode.WRONG_TOKEN_TYPE,
                    "Expected token type \"" + expectedTokenType + "\", got \""
                            + (peeked.type == null ? "" : peeked.type) + "\".");
        }

        String audience = TokenParser.peekAudience(jwt);
        if (audience == null || audience.isEmpty()) {
            throw new TokenException(TokenErrorCode.MISSING_AUDIENCE,
                    "Token is missing the `aud` (applicationAnchor) payload claim.");
        }
        String keyId = peeked.keyId;
        if (keyId == null || keyId.isEmpty()) {
            throw new TokenException(TokenErrorCode.MISSING_KEY_ID,
                    "Token is missing the `kid` header.");
        }

        JwtToken<TBody> parsed = parser.apply(jwt);

        if (!parsed.verifyExpiration(Instant.now(clock))) {
            throw new TokenException(TokenErrorCode.EXPIRED, "Token has expired.");
        }

        String publicKey;
        try {
            publicKey = resolver.resolve(audience, keyId);
        } catch (RuntimeException e) {
            throw e;
        } catch (Exception e) {
            throw new RuntimeException("PublicKeyResolver threw: " + e.getMessage(), e);
        }

        if (!parsed.verifySignature(publicKey)) {
            throw new TokenException(TokenErrorCode.INVALID_SIGNATURE,
                    "Token signature does not match the application public key.");
        }
        return parsed;
    }
}
